keps/sig-storage/1698-generic-inline-volumes/README.md

```diff
@@ -0,0 +1,540 @@
+
+When the [volume scheduling
+library](https://github.com/kubernetes/kubernetes/tree/v1.18.0/pkg/controller/volume/scheduling)
+inside the kube-scheduler or kubelet encounter such a volume inside a
+pod, they determine the name of the PVC and proceed as they currently
+do for a `PersistentVolumeClaimVolumeSource`.
+
+### User Stories
+
+#### Persistent Memory as DRAM replacement for memcached
+
+Recent releases of memcached added [support for using Persistent
+Memory](https://memcached.org/blog/persistent-memory/) (PMEM) instead
+of normal DRAM. When deploying memcached through one of the app
+controllers, `InlineVolumeSource` makes it possible to request a volume
+of a certain size from a CSI driver like
+[PMEM-CSI](https://github.com/intel/pmem-csi).
+
+#### Local LVM storage as scratch space
+
+Applications working with data sets that exceed the RAM size can
+request local storage with performance characteristics or size that is
+not met by the normal Kubernetes `EmptyDir` volumes. For example,
+[TopoLVM](https://github.com/cybozu-go/topolvm) was written for that
+purpose.
+
+### Risks and Mitigations
+
+Enabling this feature allows users to create PVCs indirectly if they can
+create pods, even if they do not have permission to create them
+directly. Cluster administrators must be made aware of this. If this
+does not fit their security model, they can disable the feature
+through the feature gate that will be added for the feature.
+
+Alternatively, a label on a namespace could be used to disable the
+feature just for that namespace.
+
+## Design Details
+
+### PVC meta data
+
+The namespace is the same as for the pod and the name is the
+concatenation of pod name and the `Volume.Name`. This is guaranteed to
+be unique for the pod and each volume in that pod. Care must be taken
+by the user to not exceed the length limit for object names, otherwise
+volumes cannot be created.
+
+Other meta data is copied from the `VolumeClaimTemplate.ObjectMeta`.
+
+### Preventing accidental collision with existing PVCs
+
+The controller will only create missing PVCs. It does not need to
+delete (handled by garbage collection) nor update (`PodSpec.Volumes`
+is immutable) existing PVCs. Therefore there is no risk that the
+controller will accidentally modify unrelated PVCs.
+
+The volume scheduling library and kubelet must check that a PVC is
+owned by the pod before using it. Otherwise they must ignored it. This
+ensures that a volume is not used accidentally for a pod in case of a
+name conflict.
+
+To surface such a name conflict to the user, the controller can also
+check for PVCs that aren't owned by the pod in its reconciliation loop
+and emit a pod event that describes the problem.
+
+### Feature gate
+
+The `GenericInlineVolumes` feature gate controls whether:
+- the new controller is active in the `kube-controller-manager`,
+- new pods can be created with a `InlineVolumeSource`,
+- kubelet and scheduler accept such volumes.
```

Comment:
Code that merely copies a `Pod.Spec.Volumes` IMHO shouldn't need changes: if it gets an `InlineVolumeSource`, it can pass that through and let those components who really have to do something special with it check the feature gate.

Let's check during the implementation phase what changes are needed where. For now I rephrased that line item a bit to be more open-ended.